### ISSUE TITLE
Users unable to click links in confirmation email

### ISSUE THREAD

**Issue by iragm** [pennersr/django-allauth#2807]
First, let me just say that I love this app, it's so simple and easy to use!

I have gotten a couple user reports of links that can't be clicked.  It seems like some apps don't properly generate a link.  At least one of the users who has reported this is using a Samsung Galaxy.

Looking at the links that work and the links that don't, it seems like there are some special characters.  Here's an example of a link that doesn't work:
https://example.com/confirm-email/NDYx:1lLQg2:IpA1jpkyIo5ZsgBqpb5lF9HC08qUxr5ophVZ3ENr2Y8/

I looked at the `EmailConfirmation` class and it seems like the class method `create` does this:
```
key = get_random_string(64).lower()
```

Is there any reason this line can't be changed to something like this?
```
key = get_random_string(64, string.ascii_lowercase+string.digits)
```

It seems like that would solve the problem because without special characters, the email app will generate the link correctly.

Alternatively, would it make sense to URI encode the key?

**Comment by 9mido** [comment]
How would you URI encode the key? 

Like this? `{{uri_goes_here|urlencode}}`

There seem to be multiple methods:

https://www.fullstackpython.com/django-utils-http-urlencode-examples.html

**Comment by iragm** [comment]
I am not super familiar with this projects code, but in the template is not a great place because most people will override the template, so this would not help many existing projects.  Also, it looks like the entire URL gets built and passed to the template.  I just tried adding a `|urlencode` to my `email_confirmation_message.txt` and it removed the URL.

I think the change to the get_random_string parameters is cleaner.  If you wanted to URI encode, I think it would need to be done in the same place, like this:

```
from django.utils.http import urlencode

# and then in the create function

def create(cls, email_address):
        key = get_random_string(64).lower()
        key = urlencode(key)
        return cls._default_manager.create(email_address=email_address, key=key)

```

**Comment by melvyn-oviavo** [comment]
First, there are no special chars in [`get_random_string()`](https://github.com/django/django/blob/stable/3.1.x/django/utils/crypto.py#L54). Second, EmailConfirmation is the old way, EmailConfirmationHMAC is the new way and the above looks like the new way, because it consists of 3 parts split by a colon.

If an email app breaks because of a colon in a URL part, then fix the email app. The time of tailoring to broken implementations has long passed. But it's also possible that your email template messes up.

Long URLs (>72 chars) in plain-text e-mails should be encapsulated by hooks:
```
<goodlink://email-reader-is-not-allowed-to-break-this-up>
```

**Comment by iragm** [comment]
Thanks for your reply, you are probably right and I am using the new way.

I can't help what apps my users choose to use, but I would like to make things as easy for them as possible.

The whole url including https//myurl.com/ and the trailing / is 90 characters.  So what you are saying is that I need <> characters around it?  If that is the case, shouldn't this file have those?  My email_confirmation_message.txt is a direct copy of this one.

https://github.com/pennersr/django-allauth/blob/master/allauth/templates/account/email/email_confirmation_message.txt

(Sorry if I am missing something obvious here)

**Comment by melvyn-oviavo** [comment]
Yes, you're correct as it's reasonable to assume the link will exceed the recommended transmission line length.

You can override the template and test it with your user.

**Comment by iragm** [comment]
I tested adding `<>` on my template and I can't say for sure if it worked or not.  I requested that the users who reported this see if they could click the link now, but surprise surprise, users who can't figure out how to copy/paste a link by themselves never got back to me.

I think it's probably safe to close this for now but I will reopen if I hear from other users that this still isn't working.  Maybe the default template should have these tags?
